an example implementation of a cnn for predicting bait performance.

yfarjoun opened the pull request:
It doesn't work very well on the ICE sample I tried it with, probably because the baits are not equimolar, and predicting the molarity is quite difficult. Still, it might be a useful template for someone.

yfarjoun committed:
an example implementation of a cnn for predicting bait performance. It doesn't work very well on the ICE sample I tried it with, probably because the baits are not equimolar, and predicting the molarity is quite difficult

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+import os
+import math
+# import h5py
```
remove unused

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+    my_metrics = [metrics.mean_squared_error, rmse_log]
+
+    model.compile(loss='mean_squared_error', optimizer=sgd, metrics=my_metrics)
+    print('model summary:\n', model.summary())
```
My bug but this should just be 
`model.summary()`
That function already contains a print

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+    # the Input layer and three Dense layers
+    model = Model(input=[input_baits, input_annotations], output=predictions)
+    model.compile(loss=gme, optimizer=sgd, metrics=my_metrics)
+    print('model summary:\n', model.summary())
```
`model.summary()`

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+                     activation="relu",
+                     init='normal'))
+
+    model.add(MaxPooling1D(pool_length=3, stride=3))
```
I try to avoid maxpooling this early in the model for genetic data.  For images we have a strong smoothness prior which we don't have on DNA sequences.

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    x = Dropout(0.2)(x)
+
+    x = MaxPooling1D(strides=3, pool_size=3)(x)
```
Maxpooling could be risky this early see above.

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    x = Flatten()(x)
+
+    x = Dense(units=32, kernel_initializer="normal", activation="relu")(x)
```
consider adding dropout here.  Typically dense layers are more prone to overfitting.  On image models people often use dropout as high as 0.5.  For vqsr our hyperparam optimizations have indicated ~0.3 is a good choice for dense dropout.

mbabadi commented:
Do we know the baits are not equimolar for a fact or just a hunch? if it is a fact, then predicting bait efficiency is obviously impossible, in particular, if the variance in molarity exceeds the context-dependent variance. The relative spatial capture efficiency of each bait, however, is not affected by this confounding factor and may be predictable.

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    xy = Dense(units=32, kernel_initializer="normal", activation="relu")(xy)
+
+    predictions = Dense(units=1, init=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation=None)(xy)
```
are predictions allowed to be negative? if not it may make sense to use a relu activation here.

yfarjoun replied:
upon rereading I now understand. you're saying: "Yes, use a RELU in the last layer."

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+            bed_with_cov_and_gc[contig][3].append(gc)
+
+    reads_per_bait = float(total_reads) / total_baits
+    print(reads_per_bait)
```
Remove or add label 
`print('reads_per_bait:', reads_per_bait)`

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+# Yossi Farjoun
+# farjoun@broadinstitute.org
+
+from __future__ import print_function
```
maybe also add:
`from __future__ import division`
This promises floating point division everywhere and allows you to use // for int division.  But it looks like there are several places where integer division is assumed so this maybe more trouble than it is worth.  I'm trying to always use it going forward as it can prevent some nasty bugs...

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    x = Dense(units=32, kernel_initializer="normal", activation="relu")(x)
+
+    print(annotation_shape)
```
remove or say what is being printed

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+                     init='normal'))
+
+    model.add(MaxPooling1D(pool_length=3, stride=3))
+    model.add(Conv1D(nb_filter=64, filter_length=16, activation="relu", init='normal', border_mode='valid'))
```
this is using the keras 1.0 argument names... stilll works but throws a warning in keras 2.0
the next model fxn uses  the keras 2.0 defaults:
`Conv1D(kernel_initializer="normal", activation="relu", padding="valid", filters=64, kernel_size=8)`

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+    bait_shape = (args.window_size, len(args.inputs),)
+    annotation_shape = (len(args.annotations),)
+
+    print(bait_shape)
```
add label or remove

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    predictions = Dense(units=1, init=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation=None)(xy)
+
+    sgd = SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True, clipnorm=0.5)
```
As we discussed maybe try the Adam optimizer.

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+    count = 0
+    while count < args.samples:
+        contig_key, start, end, coverage, gc = sample_from_bed(baits_and_coverages)
+        mid = (start + end) / 2
```
change to // if you import future division

lucidtronix commented on api_tutorials/bait_cnn/script.sh:
```diff
@@ -0,0 +1,9 @@
+grep -v '^@' whole_exome_illumina_coding_v1.Homo_sapiens_assembly19.targets.interval_list | awk 'BEGIN{FS="\t";OFS="\t"}{print $1, $2-1,$3}' > whole_exome_illumina_coding_v1.Homo_sapiens_assembly19.targets.bed
+grep -v '^@' whole_exome_illumina_coding_v1.Homo_sapiens_assembly19.baits.interval_list | awk 'BEGIN{FS="\t";OFS="\t"}{print $1, $2-1,$3}' > whole_exome_illumina_coding_v1.Homo_sapiens_assembly19.baits.bed
+
+bedtools slop -i  whole_exome_illumina_coding_v1.Homo_sapiens_assembly19.baits.bed -b 250 -g hg19.genome > sloppy.baits.bed
```
love this tool

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+    return c_idx, p_idx
+
+
+# TODO: make more random (this gives too much power to the small contigs)
```
yes!

lucidtronix reviewed (changes requested):
Cool work! Some nitpicks and suggestions...

yfarjoun committed:
changed the optimizer and the loss function. performance is somewhat improved

yfarjoun commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,528 @@
+
+    xy = Dense(units=32, kernel_initializer="normal", activation="relu")(xy)
+
+    predictions = Dense(units=1, init=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation=None)(xy)
```
confused. there is no relu here...you mean above?

yfarjoun committed:
responding to review:

- removed unused imports
- changed dropout from .2 to .3
- removed first MaxPooling layer
- added informative labels to print commands
- made division floating point

yfarjoun committed:
adding TensorBoard capability, tried to visualized the filters, but failed misrably...anyone care to see what I did wrong ?

yfarjoun committed:
added perfomance plot. network not really working, but learning a little...

yfarjoun commented:
pushed a few more commits...care to comment?

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -368,17 +386,12 @@ def split_data(data, valid_ratio=0.1, test_ratio=0.4):
+    contig_sizes = {key: len(bed_dict[key]) for key in bed_dict.keys()}
+    total_size = sum(contig_sizes.values())
 
+    contig_key = np.random.choice(bed_dict.keys(), 1, p=[x / total_size for x in contig_sizes.values()])[0]
```
Nice!

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+    lows = bed_dict[contig][0]
+    ups = bed_dict[contig][1]
+
+    return np.any((lows <= pos) & (pos <= ups))
```
Again my bug which got fixed in vqsr but never cherrypicked back here.  I believe this should be:
`return np.any((lows <= pos) & (pos < ups))`

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+
+def gme(y_true, y_pred):
+    """calculates the root (geometric) mean squared error of the values."""
+    return K.exp(K.mean(K.log(K.abs(np.divide(y_true + .001, y_pred + .001) - 1))))
```
You could use K.epsilon() instead of the hardcoded .001, this value is initialized in your ~/.keras/keras.json file and settable via K.set_epsilon(1e-05).  Also I think the division operator `/` is appropriately overloaded to handle this without np.divide()

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+    contig_sizes = {key: len(bed_dict[key]) for key in bed_dict.keys()}
+    total_size = sum(contig_sizes.values())
+
+    contig_key = np.random.choice(bed_dict.keys(), 1, p=[x / total_size for x in contig_sizes.values()])[0]
```
Nice!

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+# from math import sqrt
+#
+#
+# def put_kernels_on_grid(kernel, pad=1):
```
Can this be uncommented? Seems like a helpful fxn...

yfarjoun replied:
it might be, but I failed in getting it to work...adding it here in a comment so that folks have something to work with...

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+    model = Model(inputs=[input_baits, input_annotations], outputs=predictions)
+
+    # # add some TensorBoard annotations
+    # conv1d_1 = filter(lambda y: y.name == "conv1d_1", model.layers)[0]
```
Why is this commented out?

yfarjoun replied:
because I coudlnt' get it to work, but I wanted to show it to you to see if you could help!

lucidtronix replied:
Ok, are you trying to visualize the weights in tensorboard or something more complicated?

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+    # filters=put_kernels_on_grid(reshaped, 2)
+    #
+    # merged = tf.summary.merge_all()
+    # train_writer = tf.summary.FileWriter("./log/" + '/train')
```
Does this conflict with the TensorBoard callback?

lucidtronix commented on api_tutorials/bait_cnn/bait_performance_cnn.py:
```diff
@@ -0,0 +1,605 @@
+    # train_writer = tf.summary.FileWriter("./log/" + '/train')
+    #
+    adamo = Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=1e-08, clipnorm=1.)
+    model.compile(loss=metrics.mean_squared_error, optimizer=adamo, metrics=my_metrics)
```
I'm wondering if we should try this as a classification problem by binning the coverage first and then trying to predict (high, med, low) or (too high, high, medium, low, too low) or something.  Then we could use categorical crossentropy as our loss, which tends to have more well-behaved learning dynamics, also we could use categorical accuracy as a metric to get a quick idea of how the model compares to chance...

yfarjoun replied:
and then what?  modify the loss so that it tries to maximize xentropy + MSE?

lucidtronix replied:
I was thinking to just try categorical crossentropy though we could do it a multi-task problem where the model tries to do both.  It is easy to do with the functional API. Something like:
```
regression = Dense(units=1, kernel_initializer=RandomNormal(mean=1.0, stddev=0.5, seed=None), activation="relu")(xy)
classification = Dense(units=5, activation='softmax')(xy)
model = Model(inputs=[input_baits, input_annotations], outputs=[regression, classification])
model.compile(loss=['categorical_crossentropy', metrics.mean_squared_error], optimizer=adamo, metrics=my_metrics)
```

Assuming 5 bins for the quantized coverage...

lucidtronix reviewed (commented):
Some small changes and comment cleanup.  But the bigger question is why isn't the model learning more?  Judging by the scatter_bait_performance_cnn_model.jpg it look like *under*fitting not overfitting.  Training and test performance are similar and training error is far from 0.  Is it because there are non-sequence (and non bait position) factors that are responsible for bait coverage?  Is the equimolarity assumption wrong?  Are the bait positions normalized between 0 and 1?  They probably should be.  I would be curious how a pure sequence model performs.  Another debug idea is to cast this as classification rather than regression, see the model.compile comment on line 210.

yfarjoun commented:
I've chatted with Tim about this, he's pretty sure that the baits are NOT equi-molar...so unless I can get the molarity mixture, I'm surprised that this worked at all....

lucidtronix commented:
I'm not so surprised that there is some predictive value in the sequence alone.. we've now seen that on a few different tasks: variant filtering, indel modeling, bait performance.  Anyway, it seems like it may be difficult to track down the molarity mixture.  Should we leave this PR open while you search or do you want to merge?

yfarjoun commented:
I'll fix it up and merge...who knows how long it will take to get the molarity or spike-in list...
